frontend/controller/app/identity.js

```diff
@@ -0,0 +1,376 @@
+  if (password) {
+    const stateKeyEncryptionKeyFn = (stateEncryptionKeyId, salt) => {
+      return deriveKeyFromPassword(CURVE25519XSALSA20POLY1305, password, salt + stateEncryptionKeyId)
+    }
+
+    const { encryptionParams, value: state } = await sbp('gi.db/settings/loadEncrypted', identityContractID, stateKeyEncryptionKeyFn)
+
+    if (state) {
+      // If state contains a saved Chelonia state, extract it.
+      // The saved Chelonia state will be used to restore Chelonia on this
+      // new session.
+      const cheloniaState = state.cheloniaState
+      delete state.cheloniaState
+
+      return { encryptionParams, state, cheloniaState }
+    } else {
+      // There's no state to restore
+      return { encryptionParams, state, cheloniaState: null }
+    }
+  } else {
+    const state = await sbp('gi.db/settings/load', identityContractID)
+
+    // cheloniaState is only stored in settings when logging in with a password
+    // If there's an active session, then chelnoiaState is stored and
+    // managed separately
+    return { encryptionParams: null, state, cheloniaState: null }
+  }
+}
+
+sbp('okTurtles.events/on', LOGIN, async ({ identityContractID, encryptionParams, state }) => {
+  // This function restores the state after a successful login. It's complex
+  // because it deals with _three_ different states.
+  // * `state` (input parameter): (optional). This is the _saved_ Vuex state
+  //   that we'll use to replace the current Vuex state (`vuexState`).
+  // * `vuexState`: This is the root Vuex state as it exists. This is accessed
+  //   using the `state/vuex/state` selector. When `state` is provided, it'll
+  //   be replaced by `state`. Otherwise, it'll not be replaced.
+  // * `cheloniaState`: This is the Chelonia root state, retrieved using the
+  //   `chelonia/rootState` selector. This will be used to augment the current
+  //   or new Vuex state: any Chelonia-specific state will be set directly from
+  //   `cheloniaState` and any exisiting contract state in `state` or `vuexState`
+  //   will be discarded.
+  try {
+    const vuexState = sbp('state/vuex/state')
+    if (vuexState.loggedIn && vuexState.loggedIn.identityContractID !== identityContractID) {
+      // This shouldn't happen. It means that we received a LOGIN event but
+      // there's an active session for a different user. If this happens, it
+      // means that there's buggy login logic that should be reported and fixed
+      throw new Error('Received login event but there already is an active session')
+    }
+    const cheloniaState = cloneDeep(await sbp('chelonia/rootState'))
+    // If `state` is set, process it and replace Vuex state with it
+    if (state) {
+      // Exclude contracts from the state
+      if (state.contracts) {
+        Object.keys(state.contracts).forEach(k => {
+          // Vue.delete not needed as the entire object will replace the state
+          delete state[k]
```

Comment:
Do you mean `delete vuexState[k]` here...?

Reply:
No, I mean what's written.